src/events/SDL_categories.c

```diff
@@ -0,0 +1,226 @@
+
+    case SDL_EVENT_GAMEPAD_TOUCHPAD_DOWN:
+    case SDL_EVENT_GAMEPAD_TOUCHPAD_MOTION:
+    case SDL_EVENT_GAMEPAD_TOUCHPAD_UP:
+        return SDL_EVENTCATEGORY_GTOUCHPAD;
+
+    case SDL_EVENT_GAMEPAD_SENSOR_UPDATE:
+        return SDL_EVENTCATEGORY_GSENSOR;
+
+    case SDL_EVENT_FINGER_DOWN:
+    case SDL_EVENT_FINGER_UP:
+    case SDL_EVENT_FINGER_MOTION:
+        return SDL_EVENTCATEGORY_TFINGER;
+
+    case SDL_EVENT_CLIPBOARD_UPDATE:
+        return SDL_EVENTCATEGORY_CLIPBOARD;
+
+    case SDL_EVENT_DROP_FILE:
+    case SDL_EVENT_DROP_TEXT:
+    case SDL_EVENT_DROP_BEGIN:
+    case SDL_EVENT_DROP_COMPLETE:
+    case SDL_EVENT_DROP_POSITION:
+        return SDL_EVENTCATEGORY_DROP;
+
+    case SDL_EVENT_AUDIO_DEVICE_ADDED:
+    case SDL_EVENT_AUDIO_DEVICE_REMOVED:
+    case SDL_EVENT_AUDIO_DEVICE_FORMAT_CHANGED:
+        return SDL_EVENTCATEGORY_ADEVICE;
+
+    case SDL_EVENT_SENSOR_UPDATE:
+        return SDL_EVENTCATEGORY_SENSOR;
+
+    case SDL_EVENT_PEN_DOWN:
+    case SDL_EVENT_PEN_UP:
+        return SDL_EVENTCATEGORY_PTIP;
+
+    case SDL_EVENT_PEN_MOTION:
+        return SDL_EVENTCATEGORY_PMOTION;
+
+    case SDL_EVENT_PEN_BUTTON_DOWN:
+    case SDL_EVENT_PEN_BUTTON_UP:
+        return SDL_EVENTCATEGORY_PBUTTON;
+
+    case SDL_EVENT_CAMERA_DEVICE_ADDED:
+    case SDL_EVENT_CAMERA_DEVICE_REMOVED:
+    case SDL_EVENT_CAMERA_DEVICE_APPROVED:
+    case SDL_EVENT_CAMERA_DEVICE_DENIED:
+        return SDL_EVENTCATEGORY_CDEVICE;
+    }
+}
+
+SDL_Window *SDL_GetWindowFromEvent(const SDL_Event *event)
+{
+    SDL_WindowID windowID;
+
+    switch (SDL_GetEventCategory(event->type)) {
+    case SDL_EVENTCATEGORY_USER:
+        windowID = event->user.windowID;
+        break;
+    case SDL_EVENTCATEGORY_WINDOW:
+        windowID = event->window.windowID;
+        break;
+    case SDL_EVENTCATEGORY_KEY:
+        windowID = event->key.windowID;
+        break;
+    case SDL_EVENTCATEGORY_EDIT:
+        windowID = event->edit.windowID;
+        break;
+    case SDL_EVENTCATEGORY_TEXT:
+        windowID = event->text.windowID;
+        break;
+    case SDL_EVENTCATEGORY_EDIT_CANDIDATES:
+        windowID = event->edit_candidates.windowID;
+        break;
+    case SDL_EVENTCATEGORY_MOTION:
+        windowID = event->motion.windowID;
+        break;
+    case SDL_EVENTCATEGORY_BUTTON:
+        windowID = event->button.windowID;
+        break;
+    case SDL_EVENTCATEGORY_WHEEL:
+        windowID = event->wheel.windowID;
+        break;
+    case SDL_EVENTCATEGORY_TFINGER:
+        windowID = event->tfinger.windowID;
+        break;
+    case SDL_EVENTCATEGORY_PTIP:
+        windowID = event->ptip.windowID;
+        break;
+    case SDL_EVENTCATEGORY_PMOTION:
+        windowID = event->pmotion.windowID;
+        break;
+    case SDL_EVENTCATEGORY_PBUTTON:
+        windowID = event->pbutton.windowID;
+        break;
+    case SDL_EVENTCATEGORY_DROP:
+        windowID = event->drop.windowID;
+        break;
+    default:
+        return NULL;
```

Comment:
We need to call SDL_SetError() here.

Reply:
This is not an error condition.
Event categories not listed have no associated window.
For invalid event types, `SDL_GetEventCategory` will set an error message.

Reply:
Perhaps I should add a `-1` case to make this clear in code (with comments).